### PR TITLE
Add timeouts to unary client

### PR DESCRIPTION
**Pull request by rchowinfoblox**
All UTs and benchmarks (make build test bench) pass successfully.

### CODE REVIEW

File: pep/unary_client.go [hunk 1]
```diff
@@ -84,7 +84,16 @@ func (c *unaryClient) Connect(addr string) error {
 		return err
 	}
 
-	conn, err := grpc.Dial(addr, opts...)
+	var conn *grpc.ClientConn
+	if c.opts.connTimeout > 0 {
+		opts = append(opts, grpc.WithBlock())
+		toCtx, toCancelFn := context.WithTimeout(context.Background(), c.opts.connTimeout)
+		defer toCancelFn()
+		conn, err = grpc.DialContext(toCtx, addr, opts...)
```
Comment (rchowinfoblox) [inline pep/unary_client.go]: Note that with the timeout, Connect() becomes a blocking call.  Without the timeout, Connect() remains a non-blocking call.

**Comment by rchowinfoblox** [comment]
I don't know why TravisCI is failing on fmt-pep, when I run it locally it, fmt-pep succeeds

**Comment by drewwells** [comment]
> I don't know why TravisCI is failing on fmt-pep, when I run it locally it, fmt-pep succeeds

Connect() is always a blocking call, that's why this PR is critical

File: pep/unary_client.go [hunk 2]
```diff
@@ -84,7 +84,21 @@ func (c *unaryClient) Connect(addr string) error {
 		return err
 	}
 
-	conn, err := grpc.Dial(addr, opts...)
+	ctx := c.opts.ctx
+	if ctx == nil {
+		ctx = context.Background()
+	}
```
[the remark this answers is not in the document]
Comment (rchowinfoblox) [reply pep/unary_client.go]: client is now init with default nil context

File: pep/unary_client.go [hunk 3]
```diff
@@ -84,7 +84,21 @@ func (c *unaryClient) Connect(addr string) error {
...
+
+	var conn *grpc.ClientConn
+	if c.opts.connTimeout > 0 {
+		opts = append(opts, grpc.WithBlock())
+		toCtx, toCancelFn := context.WithTimeout(ctx, c.opts.connTimeout)
+		defer toCancelFn()
+		conn, err = grpc.DialContext(toCtx, addr, opts...)
+	} else {
+		conn, err = grpc.Dial(addr, opts...)
+	}
```
[the remark this answers is not in the document]
Comment (rchowinfoblox) [reply pep/unary_client.go]: done

File: pep/client.go [hunk 4]
```diff
@@ -218,6 +231,7 @@ type options struct {
 // NewClient creates client instance using given options.
 func NewClient(opts ...Option) Client {
 	o := options{
+		ctx:            context.Background(),
```
[the remark this answers is not in the document]
Comment (rchowinfoblox) [reply pep/client.go]: Removed context init, which defaults to nil

File: pep/streaming_connection.go [hunk 5]
```diff
@@ -197,7 +199,12 @@ func (c *streamConn) tryConnect(addr string, tracer opentracing.Tracer, timeout
 		)
 	}
 
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	ctx := c.ctx
+	if ctx == nil {
+		ctx = context.Background()
+	}
```
[the remark this answers is not in the document]
Comment (rchowinfoblox) [reply pep/streaming_connection.go]: client is now init with default nil context

File: pep/unary_client.go [hunk 6]
```diff
@@ -162,7 +176,19 @@ func (c *unaryClient) Validate(in, out interface{}) error {
 		}
 	}
 
-	res, err := (*uc).Validate(context.Background(), &req, grpc.FailFast(false))
+	ctx := c.opts.ctx
+	if ctx == nil {
+		ctx = context.Background()
+	}
```
[the remark this answers is not in the document]
Comment (rchowinfoblox) [reply pep/unary_client.go]: client is now init with default nil context

**Pull request merged**